### ISSUE TITLE
create a network resource

### ISSUE THREAD

**Issue by purpleidea** [purpleidea/mgmt#28]
Managing a network interface is a fundamental part of managing any system. We should provide a native interface to do so. Some ideas:

* It might be worth basing this on systemd-networkd. (It looks like a safe, stable base!)
* Since mgmt works in a decentralized manner, so transient network "blips" shouldn't disrupt it.
* During CheckApply() if we don't transition to an acceptable state, we could revert and error.
* With the powerful Resource (Res) API, other resources will be able to add [automatic edges](https://ttboj.wordpress.com/2016/03/14/automatic-edges-in-mgmt/) to this network resource in case they require connectivity. Yet another reason this is an important primitive to write.
* I'd focus on getting simple networking working initially, with support for bridges or other systemd-networkd features as an exciting second step.

Patches welcome! Please feel free to make yourself known if you'd like to work on this

**Comment by mikefaille** [comment]
Maybe, we can simply build plugin system as Docker's plugins or Ansible module. The goal is to provide way to implement many solutions. 

At least, we can enforce common api for all systems. And, network management solutions can exist in many ways like : 
- systemd-networkd
- network-manager (used for desktop)
- systems without network manager activated for one specific interface. In this case, we can directly implement netlink's kernel interface used by libcontainer --> https://github.com/mikefaille/tenus/blob/master/macvtap_linux.go
- etc

**Comment by purpleidea** [comment]
@mikefaille I don't think we need a plugin system, at least not until we have more core base resources.

> systemd-networkd

Indeed this is mentioned in the original description.

> systems without network manager activated for one specific interface

Does this code support the interfaces to generating events when something changes? I expect it will be best to wrap networkd for now.

These issue tickets are open to attract someone to want to work on the code. Since you're commenting, are you volunteering? :)

**Comment by LrsK** [comment]
As I was prodded by @purpleidea, I'm taking a stab at this! So far these are my thoughts on how the resource should work.

The resource controls the systemd-networkd configuration, which takes place via configuration files that are placed in the directory /etc/systemd/network. Networks are configured in files with extension ".network", and virtual network devices are configured in files with extension ".netdev". So the mgmt resource will describe and enforce one or more network configurations. Each configuration ends up in a file e.g.  `/etc/systemd/network/10-static_ip.network`

Configuration files are sorted and processed in lexical order by the networkd service on startup. The first file that has a `[Match]` field that matches one of its devices among all the config files is used, even if a later file matches as well, so priority matters. The resource should let the user describe the configuration with all the same options that exist in networkd, where several validations will have to take place, like IP adresses, CIDRs, MAC addresses, as well as some conditional cases. Finally the service will have to be restarted for the changes to take effect.

This will result in one or more config files, but some files may already exist on the system. Question then is what should happen to the existing config? Should all existing files be deleted, and replaced by new ones defined by the user? What if something goes wrong in the new config? You could lose contact with the machine. Perhaps a backup can be taken first, and roll back somehow upon some test failing.

**Comment by purpleidea** [comment]
@LrsK Hey, feel free to join #mgmtconfig on Freenode IRC if you'd prefer a more back and forth conversation, also ghangouts are possible. Here are a few replies:

1) Only devices which we decide to manage get managed. Everything else should be ignored. If the user tries to do a badly planned mixed management approach, then it's really their fault.

2) Rollback is something I'm especially excited about. One of the goals of mgmt being distributed is that the net resource could have a sort of health check function where if the new config didn't satisfy this check (and it can't find it's peers anymore) we can rollback to the old config and generate an error. I've got more ideas here if you'd like to talk more.

3) We can use the existing file resource code to make this work as a composite resource, so no need to rewrite all of it.

4) We might need to get netlink events somehow to know when an interface changes.

LMK if you have more questions!

**Comment by LrsK** [comment]
Something else I was thinking about was that networkd isn't used that much yet, at least not in some of the distros I've tested on like Ubuntu 16.04, Fedora 24, CentOS. It is indeed possible to use networkd on these systems, but they don't seem to use them by default.

Would it be a good idea to allow the user to specify what format they would like to output the config files in, and place them in the appropriate place on the system? We could solve this by taking in the set options from the yaml file like normal, and then use text/template to produce the actual files. This way, mgmt can even be used on systems without systemd like Alpine or Gentoo.

This would of course be a bigger job in total, but perhaps we should implement networkd support with this flexibility in mind?

**Comment by mikefaille** [comment]
Personally, I don't recommend to bind stuff on least use tech only for coolnest. To use networkd, you must prepare your OS to transfert network configuration (includinding pre/post script).

**Comment by purpleidea** [comment]
@LrsK I think it is a good idea to support networkd initially because it is IMO the most promising future for server networking. If down the road something else turns out to be more useful, we can always implement that as a second resource.

**Comment by sadegh-nikaein** [comment]
Check this 
https://github.com/milosgajdos83/tenus

**Comment by mikefaille** [comment]
@purpleidea I personnally prefer anostic solution like Tenus. I tried it in the past : https://github.com/mikefaille/docker-group-daemonizer/blob/master/main.go

**Comment by sadegh-nikaein** [comment]
for interface address change events  we can use net/route package wait until  for sysRTM_NEWADDR & sysRTM_DELADDR message trigger

**Comment by purpleidea** [comment]
@sadegh-nikaein Want to either write the base resource or a small proof of concept standalone mainloop showing the events happening?

**Comment by sadegh-nikaein** [comment]
@purpleidea i try that is not work as i expected .
what about this idea etcd package netutil 
https://github.com/coreos/etcd/blob/master/pkg/netutil/routes_linux.go
has a function call GetDefaultHost return the first IP address of machine from the routing table
we can store it somewhere and check it with some interval or 
we can on init add bridge to that with Tenus and not matter interface change or not  
or maybe mix of both !?

**Comment by purpleidea** [comment]
@sadegh-nikaein Getting/Setting on the interface is the easy part, the less straight forward part is getting events, eg from NETLINK. In any case it should be a straight forward exercise for someone wanting to do the patching. LMK if you're interested.

**Comment by sadegh-nikaein** [comment]
I try my best but its higher than my lvl i more research and find good example to get status with NETLINK and this is possible take look at this 
https://github.com/golang/go/issues/15080
https://github.com/MichielDeMey/home-controller/blob/d26d5f131856db6b34fff3dc2853c0c760c231f8/lib/monitor.go
https://github.com/inoyatov/hipp/blob/master/main.go
https://github.com/vishvananda/netlink/blob/master/addr_linux.go#L222
